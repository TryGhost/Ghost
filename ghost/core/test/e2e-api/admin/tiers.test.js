const assert = require('assert/strict');
const {
    agentProvider,
    fixtureManager,
    mockManager,
    matchers
} = require('../../utils/e2e-framework');
const models = require('../../../core/server/models/index');
const {anyContentVersion, anyEtag} = matchers;

describe('Tiers API', function () {
    let agent;

    before(async function () {
        agent = await agentProvider.getAdminAPIAgent();
        await fixtureManager.init('members');
        await agent.loginAsOwner();
    });

    afterEach(function () {
        mockManager.restore();
    });

    it('Can browse Tiers', async function () {
        await agent
            .get('/tiers/')
            .expectStatus(200)
            .matchBodySnapshot({
                tiers: Array(2).fill({
                    id: matchers.anyObjectId,
                    created_at: matchers.anyISODateTime,
                    updated_at: matchers.anyISODateTime
                })
            });
    });

    it('Errors when price is non-integer', async function () {
        const tier = {
            name: 'Blah',
            monthly_price: 99.99,
            currency: 'usd'
        };

        await agent
            .post('/tiers/')
            .body({tiers: [tier]})
            .expectStatus(422)
            .matchBodySnapshot({
                errors: [{
                    id: matchers.anyUuid
                }]
            });
    });

    it('Errors when price is negative', async function () {
        const tier = {
            name: 'Blah',
            monthly_price: -100,
            currency: 'usd'
        };

        await agent
            .post('/tiers/')
            .body({tiers: [tier]})
            .expectStatus(422)
            .matchBodySnapshot({
                errors: [{
                    id: matchers.anyUuid
                }]
            });
    });

    it('Errors when price is too large', async function () {
        const tier = {
            name: 'Blah',
            monthly_price: Number.MAX_SAFE_INTEGER,
            currency: 'usd'
        };

        await agent
            .post('/tiers/')
            .body({tiers: [tier]})
            .expectStatus(422)
            .matchBodySnapshot({
                errors: [{
                    id: matchers.anyUuid
                }]
            });
    });

    it('Can read Tiers', async function () {
        const {body: {tiers: [tier]}} = await agent.get('/tiers/');

        await agent.get(`/tiers/${tier.id}/`)
            .expectStatus(200);
    });

    it('Can edit visibility', async function () {
        const {body: {tiers: [tier]}} = await agent.get('/tiers/?filter=type:paid&limit=1');

        const visibility = tier.visibility === 'none' ? 'public' : 'none';

        await agent.put(`/tiers/${tier.id}/`)
            .body({
                tiers: [{
                    visibility
                }]
            })
            .expectStatus(200);

        const {body: {tiers: [updatedTier]}} = await agent.get(`/tiers/${tier.id}/`);

        assert(updatedTier.visibility === visibility, `The visibility of the Tier should have been updated to ${visibility}`);
    });

    it('Can save with trial_days as null', async function () {
        const {body: {tiers: [tier]}} = await agent.get('/tiers/?limit=1');

        await agent.put(`/tiers/${tier.id}/`)
            .body({
                tiers: [{
                    trial_days: null
                }]
            })
            .expectStatus(200);

        const {body: {tiers: [updatedTier]}} = await agent.get(`/tiers/${tier.id}/`);

        assert(updatedTier.trial_days === 0, `The trial_days should have been set to 0`);
    });

    it('Can edit tier properties and relations', async function () {
        let {body: {tiers: [tier]}} = await agent.get('/tiers/?filter=type:paid&limit=1')
            .expectStatus(200)
            .matchHeaderSnapshot({
                'content-version': anyContentVersion,
                etag: anyEtag
            })
            .matchBodySnapshot({
                tiers: Array(1).fill({
                    id: matchers.anyObjectId,
                    created_at: matchers.anyISODateTime,
                    updated_at: matchers.anyISODateTime
                })
            });

        await agent.put(`/tiers/${tier.id}/`)
            .body({
                tiers: [{
                    description: 'Updated description',
                    benefits: ['daily cat pictures', 'delicious avo toast']
                }]
            })
            .expectStatus(200);

        await agent.get(`/tiers/${tier.id}/`)
            .expectStatus(200)
            .matchHeaderSnapshot({
                'content-version': anyContentVersion,
                etag: anyEtag
            })
            .matchBodySnapshot({
                tiers: Array(1).fill({
                    id: matchers.anyObjectId,
                    created_at: matchers.anyISODateTime,
                    updated_at: matchers.anyISODateTime
                })
            });
    });

    it('Can create a new tier', async function () {
        const tier = {
            name: 'new premium tier',
            monthly_price: 100,
            currency: 'usd'
        };

        await agent
            .post('/tiers/')
            .body({tiers: [tier]})
            .expectStatus(201)
            .matchBodySnapshot({
                tiers: Array(1).fill({
                    id: matchers.anyObjectId,
                    created_at: matchers.anyISODate,
                    name: 'new premium tier',
                    slug: 'new-premium-tier',
                    monthly_price: 100,
                    currency: 'USD'
                })
            });
    });

    it('Can update a benefit\'s capitalization', async function () {
        const tierData = {
            name: 'benefit test tier',
            monthly_price: 100,
            currency: 'usd',
            benefits: ['TEST BENEFIT']
        };

        let {body: {tiers: [tier]}} = await agent.post('/tiers/')
            .body({tiers: [tierData]})
            .expectStatus(201)
            .matchBodySnapshot({
                tiers: Array(1).fill({
                    id: matchers.anyObjectId,
                    created_at: matchers.anyISODate,
                    name: 'benefit test tier',
                    slug: 'benefit-test-tier',
                    monthly_price: 100,
                    currency: 'USD'
                })
            });

        await agent.put(`/tiers/${tier.id}/`)
            .body({
                tiers: [{
                    benefits: ['Test benefit']
                }]
            })
            .expectStatus(200);

        await agent.get(`/tiers/${tier.id}/`)
            .expectStatus(200)
            .matchHeaderSnapshot({
                'content-version': anyContentVersion,
                etag: anyEtag
            })
            .matchBodySnapshot({
                tiers: Array(1).fill({
                    id: matchers.anyObjectId,
                    created_at: matchers.anyISODate,
                    benefits: ['Test benefit']
                })
            });

        const benefit = await models.Benefit.findOne({slug: 'test-benefit'});
        assert(benefit.attributes.name === 'Test benefit', 'The benefit should have been updated.');
        const previousBenefit = await models.Benefit.findOne({slug: 'test-benefit-2'});
        assert(!previousBenefit, 'The previous benefit should have been overwritten');
    });
});
