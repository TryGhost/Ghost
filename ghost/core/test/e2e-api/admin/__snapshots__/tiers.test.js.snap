// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tiers API Can browse Tiers 1: [body] 1`] = `
Object {
  "meta": Object {
    "pagination": Object {
      "limit": 2,
      "next": null,
      "page": 1,
      "pages": 1,
      "prev": null,
      "total": 2,
    },
  },
  "tiers": Array [
    Object {
      "active": true,
      "benefits": Array [],
      "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "description": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "name": "Free",
      "slug": "free",
      "trial_days": 0,
      "type": "free",
      "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "visibility": "public",
      "welcome_page_url": "/welcome-free",
    },
    Object {
      "active": true,
      "benefits": Array [],
      "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "currency": "USD",
      "description": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "monthly_price": 500,
      "name": "Default Product",
      "slug": "default-product",
      "trial_days": 0,
      "type": "paid",
      "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "visibility": "public",
      "welcome_page_url": "/welcome-paid",
      "yearly_price": 5000,
    },
  ],
}
`;

exports[`Tiers API Can create a new tier 1: [body] 1`] = `
Object {
  "tiers": Array [
    Object {
      "active": true,
      "benefits": Array [],
      "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}/,
      "currency": "USD",
      "description": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "monthly_price": 100,
      "name": "new premium tier",
      "slug": "new-premium-tier",
      "trial_days": 0,
      "type": "paid",
      "updated_at": null,
      "visibility": "public",
      "welcome_page_url": null,
      "yearly_price": 5000,
    },
  ],
}
`;

exports[`Tiers API Can edit tier properties and relations 1: [body] 1`] = `
Object {
  "meta": Object {
    "pagination": Object {
      "limit": 1,
      "next": null,
      "page": 1,
      "pages": 1,
      "prev": null,
      "total": 1,
    },
  },
  "tiers": Array [
    Object {
      "active": true,
      "benefits": Array [],
      "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "currency": "USD",
      "description": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "monthly_price": 500,
      "name": "Default Product",
      "slug": "default-product",
      "trial_days": 0,
      "type": "paid",
      "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "visibility": "none",
      "welcome_page_url": "/welcome-paid",
      "yearly_price": 5000,
    },
  ],
}
`;

exports[`Tiers API Can edit tier properties and relations 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "450",
  "content-type": "application/json; charset=utf-8",
  "content-version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Version, Origin, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Tiers API Can edit tier properties and relations 3: [body] 1`] = `
Object {
  "tiers": Array [
    Object {
      "active": true,
      "benefits": Array [
        "daily cat pictures",
        "delicious avo toast",
      ],
      "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "currency": "USD",
      "description": "Updated description",
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "monthly_price": 500,
      "name": "Default Product",
      "slug": "default-product",
      "trial_days": 0,
      "type": "paid",
      "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "visibility": "none",
      "welcome_page_url": "/welcome-paid",
      "yearly_price": 5000,
    },
  ],
}
`;

exports[`Tiers API Can edit tier properties and relations 4: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "422",
  "content-type": "application/json; charset=utf-8",
  "content-version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Version, Origin, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Tiers API Can update a benefit's capitalization 1: [body] 1`] = `
Object {
  "tiers": Array [
    Object {
      "active": true,
      "benefits": Array [
        "TEST BENEFIT",
      ],
      "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}/,
      "currency": "USD",
      "description": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "monthly_price": 100,
      "name": "benefit test tier",
      "slug": "benefit-test-tier",
      "trial_days": 0,
      "type": "paid",
      "updated_at": null,
      "visibility": "public",
      "welcome_page_url": null,
      "yearly_price": 5000,
    },
  ],
}
`;

exports[`Tiers API Can update a benefit's capitalization 1: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "832",
  "content-type": "application/json; charset=utf-8",
  "content-version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Version, Origin, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Tiers API Can update a benefit's capitalization 2: [body] 1`] = `
Object {
  "tiers": Array [
    Object {
      "active": true,
      "benefits": Array [
        "Test benefit",
      ],
      "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}/,
      "currency": "USD",
      "description": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "monthly_price": 100,
      "name": "benefit test tier",
      "slug": "benefit-test-tier",
      "trial_days": 0,
      "type": "paid",
      "updated_at": null,
      "visibility": "public",
      "welcome_page_url": null,
      "yearly_price": 5000,
    },
  ],
}
`;

exports[`Tiers API Can update a benefit's capitalization 3: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "350",
  "content-type": "application/json; charset=utf-8",
  "content-version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Version, Origin, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Tiers API Errors when price is negative 1: [body] 1`] = `
Object {
  "errors": Array [
    Object {
      "code": null,
      "context": "Tier prices must not be negative",
      "details": null,
      "ghostErrorCode": null,
      "help": null,
      "id": StringMatching /\\[a-f0-9\\]\\{8\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{12\\}/,
      "message": "Validation error, cannot save tier.",
      "property": null,
      "type": "ValidationError",
    },
  ],
}
`;

exports[`Tiers API Errors when price is non-integer 1: [body] 1`] = `
Object {
  "errors": Array [
    Object {
      "code": null,
      "context": "Tier prices must be an integer.",
      "details": null,
      "ghostErrorCode": null,
      "help": null,
      "id": StringMatching /\\[a-f0-9\\]\\{8\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{12\\}/,
      "message": "Validation error, cannot save tier.",
      "property": null,
      "type": "ValidationError",
    },
  ],
}
`;

exports[`Tiers API Errors when price is too large 1: [body] 1`] = `
Object {
  "errors": Array [
    Object {
      "code": null,
      "context": "Tier prices may not exceed 999999.99",
      "details": null,
      "ghostErrorCode": null,
      "help": null,
      "id": StringMatching /\\[a-f0-9\\]\\{8\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{12\\}/,
      "message": "Validation error, cannot save tier.",
      "property": null,
      "type": "ValidationError",
    },
  ],
}
`;
